Dots-Per-Inch (DPI) in Plotter Module

wbinventor opened the pull request:
This PR adds a new `dpi` parameter to all of the routines in the `openmoc.plotter` module. The default matplotlib dpi is 100 and this is now over-ridden to 500 as an OpenMOC default, but this can be specified by the user as well. In addition, the OpenMC MGXS `Libraries` for the SPH factor sample inputs (slab and pin cell) have been updated to conform to the latest version of OpenMC.

wbinventor commented:
My latest commit to this PR goes one step further and adds in support for the Python Imaging Library (PIL) in addition to matplotlib. We can now save our figures as pure bitmaps using PIL which look *much* better than the images from matplotlib. This is particularly true for high-res figures in which we may set the `gridsize` parameter to 1000 or more. In these cases, matplotlib's `imshow` method will provide diminishing returns since the resolution is only improved without first setting the figure size. Although this is easy to do, I think using PIL directly is a better option since it automatically scales with `gridsize` (the number of pixels) and does not require a new input parameter for `figsize`.

One can choose between `'matplotlib'` and `'pil'` using the new optional `library` parameter to most of the routines in `openmoc.plotter`.  Of course matplotlib has a lot of features that PIL does not, including colorbars, scales for axes, labels, etc., so `'matplotlib'` is kept as the default library.

wbinventor commented:
Just as a reference point, using PIL allows for crisp plots of complicated geometries such as the following spectrally clustered PWR geometries. These plots look like pixelated crap when generated with `matplotlib.plt.imshow`.

**2x2 BEAVRS Assemblies w/ Grid Spacers**
<img src="https://cloud.githubusercontent.com/assets/209492/13029775/9e25bcb2-d263-11e5-917e-9de40ab7bed9.png" width=350>

**2x2 BEAVRS Assemblies w/ Grid Spacer and Reflector**
<img src="https://cloud.githubusercontent.com/assets/209492/13029756/f4be54ea-d262-11e5-8f1c-e294472de150.png" width=350>

Plots such as these are easily generated and displayed inline in Jupyter Notebooks as follows:

```python
openmoc.plotter.plot_materials(geometry, gridsize=1000, library='pil')
IPython.display.Image(filename='plots/materials.png', width=400, height=400)
```

samuelshaner commented:
Wow - those plots are beautiful!

samuelshaner commented on openmoc/plotter.py:
```diff
@@ -414,7 +430,7 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None,
 
     # Plot centroids on top of 2D flat source region color map
-    if centroids:
+    if library == 'matplotlib' and centroids:
```
Can the centroids not be plotted with `pil`?

wbinventor commented:
Now you can :-) I'm going to add some tests to the test suite to cover a lot of our plotting functionality. Once that is in, this PR will address everything I set out to do. But the new plotting features - DPI and PIL - can be reviewed independently of the new tests.

wbinventor commented:
Alright, I think this PR is ready for a full review. My last series of commits do the following:

* Added 9 new tests to suite for near full coverage of `openmoc.plotter`
* Fixed `update_results.py` to first install OpenMOC locally just as is done in `run_tests.py`
* Fixed test suite issues with Python 3
* Removed `dpi` param from `openmoc.plotter` routines since PIL is more effective way to produce high-res images. Plus, one can still adjust the DPI in `matplotlib_rcparams`.

Note that the new plotting tests only work with Matplotlib >=v1.5. Older versions generate plots with slightly different pixel counts, which fail the tests when those images are hashed and compared to the true hashed results generated with >=v1.5.

samuelshaner commented on openmoc/plotter.py:
```diff
@@ -1028,48 +1074,66 @@ def plot_spatial_data(domains_to_data, plot_params, get_figure=False):
-        if get_figure:
-            figures.append(fig)
+        # Append file extension (e.g., '.png', '.ppm') to filename
+        plot_filename +=  plot_params.extension
```
extra spacing after `+=`

samuelshaner commented on openmoc/plotter.py:
```diff
@@ -540,10 +581,11 @@ def plot_spatial_fluxes(solver, energy_groups=[1], norm=False,
     plot_params.ylim = ylim
     plot_params.colorbar = True
-    plot_params.cmap = None
+    plot_params.cmap = plt.get_cmap('jet')
```
Is there any reason the default should not just be set to `jet`?

wbinventor replied:
We use matplotlib's default colormap for all plots except the flux and fission rates (or if one specifies one to `PlotParams` for `plot_spatial_data(...)`). Because of this I decided to keep the default as `None` and only specify `jet` for the two plot types where we want it.

samuelshaner commented on tests/testing_harness.py:
```diff
@@ -258,3 +263,57 @@ def _get_results(self, num_iters=False, keff=False, fluxes=False,
+    def __init__(self):
+        super(PlottingTestHarness, self).__init__()
+        self.figures = []
+
```
I think PEP 8 suggests only one blank line between class methods.

samuelshaner commented on tests/testing_harness.py:
```diff
@@ -258,3 +263,57 @@ def _get_results(self, num_iters=False, keff=False, fluxes=False,
+            # Hash the image and append to output string
+#            sha512 = hashlib.sha512()
+#            sha512.update(outstr.encode('utf-8'))
+#            outstr = sha512.hexdigest()
```
I'm assuming these commented out lines should be removed?

wbinventor replied:
Yes

samuelshaner commented:
This PR looks great! I really like the new crisper plots. I'm happy with it once my comments are addressed.

wbinventor commented:
Just addressed all of your comments @samuelshaner.

geogunow commented:
Looks good to me!
